### DISCUSSION TITLE
StrongSwan VPN with Custom Root CA Cert

### DISCUSSION THREAD

**Discussion by 0-SamboNZ-0** [strongswan/strongswan#1978] [Q&A] [answered]
Can anyone direct me to some instructions for installing a custom Root CA Cert on my Galaxy S7 (Android 8)?

I've managed to install it, but it shows up as a user cert rather than a Root and my VPN client needs to see it as a root cert AFAICS.

VPN connection fails and I think this is why.

**Comment by 0-SamboNZ-0** [comment]
So after consulting with some Android experts, it seems that post Android 6, root access is required to install Root CAs.

Additionally, the consensus is that the cert shouldn't have to be installed as a Root CA.

The secondary issue here is that the StrongSwan VPN client can't see the cert I've installed, and when I try to install the cert using the StrongSwan client, it tells me that it "Failed to import certificate".

I've tried .P12 and .PEM formats, with and without passwords.

**Reply by tobiasbrunner** [reply]
> So after consulting with some Android experts, it seems that post Android 6, root access is required to install Root CAs.

Not sure what you are referring to, but at least on a Pixel phone it's definitely possible to install CA certificates with newer Android versions. There is a warning about it, and I wouldn't recommend it, but it's still possible.





> The secondary issue here is that the StrongSwan VPN client can't see the cert I've installed, and when I try to install the cert using the StrongSwan client, it tells me that it "Failed to import certificate".

Then the file does not contain a valid certificate. If it's a PKCS#12 container, then that won't work. You have to extract the CA certificate from it and  install it separately. But if you imported the PKCS#12 file into the system to use the contained client certificate/key, that should also have imported the CA certificate of that chain if it was contained in the file. That should work if the automatic CA selection is enabled. If you want to select the CA certificate specifically, you have to extract and import it separately.

**Reply by 0-SamboNZ-0** [reply]
Hi @tobiasbrunner,

So I successfully installed the certificate onto my Android device via the native certificate import method.
It now shows up in 'User Certificates' and contains:
- 1 user key
- 1 user certificate
- 1 CA certificate

However, it shows as 'installed for Wi-Fi', not 'Installed for VPN'
I can't see a way to change that.

**Reply by tobiasbrunner** [reply]
In the system settings there is an option to install a "VPN & app user certificate" or a "Wi-Fi certificate". Whether it makes a difference, I don't know. If it shows up via KeyChain API in the strongSwan app, it doesn't.

**Comment by 0-SamboNZ-0** [comment]
> In the system settings there is an option to install a "VPN & app user certificate" or a "Wi-Fi certificate".

Hmm, I can find no such option on my device.

**Reply by tobiasbrunner** [reply]
Might be different for Samsung devices, no idea. But it's pretty well hidden deep down in the security settings anyway, so searching for it might be the easiest way to get to it.

**Reply by 0-SamboNZ-0** [reply]
Ok, I discovered that there are actually 2 ways to install a cert from my settings:
- Connections / Wifi / Advanced (which is what I originally used, which probably explains the result)
- Lock Screen and Security / Other Security Settings / Install from Device Storage (Install Certificates from Storage)

Installing from the second location gives me the option to select 'Credentials Use' as 'VPN and Apps' or 'Wi-Fi'.

Notes:
- The certificate file must be in the 'Downloads' folder
- When I try to run the second option but from settings search, I get the error 'No Certificates to Install', so that's a bit misleading.

HOWEVER; the cert still doesn't show in the StrongSwan VPN client and I still can't connect to my VPN :)

**Reply by tobiasbrunner** [reply]
Where doesn't it show up? When you try to select it as client certificate in the profile settings? Or in the list of CA certificates? (It won't show up in the latter. The CA certificate is only loaded together with the client certificate when the connection is getting established.) Also check the log in the app for details on why it failed to connect.

**Reply by 0-SamboNZ-0** [reply, marked as answer]
It didn't show up in the list of certs and trying to manually select it in the profile settings just took me to the list (without the cert showing).

In any case, I figured it out.

In the end, importing the cert into the Android system wasn't required.

What WAS required was the following:
- Export the PKCS#12 cert package from the VPN server
- Open the PKCS#12 cert package & export the CA root and Server certs from the package (I saved as a PKCS#7 package, but I suppose you could individually export as .CER files and get the same result?) - [Performed on a Windows 10 PC]
- Within the StrongSwan VPN client, go to the 'CA Certificates' screen
- Select 'Install Certificate'
- Select the PKCS#7 cert package file

Once I did this, the VPN Server cert showed up in the CA Certificates screen under 'Imported' and I could connect to my VPN!

Thanks for you assistance with this @tobiasbrunner, I really appreciate it!
